Read count logging for PathSeq Filter, Score, and Pipeline tools

mwalker174 opened the pull request:
Read counts at different stages of the PathSeq pipeline are now logged using `MetricsFile`. The filter metrics contains the number of reads remaining and number of reads filtered at each step (after filtering pre-aligned reads, low quality/complexity reads, host reads, and duplicates). The score metrics give number of pathogen-mapped and unmapped reads.

These metrics are now validated in the PathSeq integration tests, which have also been refactored to use DataProviders instead of separate functions.

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -29,9 +29,9 @@
 public class PSFilterTest extends CommandLineProgramTest {
 
-    private final String BWA_IMAGE_PATH = "src/test/resources/" + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.bwa_image";
+    private final String BWA_IMAGE_PATH = "src/test/resources/" + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.img";
```
camelUpperCase or make it static.

mwalker174 replied:
Made static

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -29,9 +29,9 @@
-    private final String BWA_IMAGE_PATH = "src/test/resources/" + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.bwa_image";
+    private final String BWA_IMAGE_PATH = "src/test/resources/" + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.img";
 
-    final static List<List<String>> readFastq(final File fastqFile) {
```
Could you use SVFastqUtils method to deal with Fastq formatted files? Otherwise is there something in htsjdk to handle Fastq?

mwalker174 replied:
Now using htsjdk fastqreader

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSparkIntegrationTest.java:
```diff
@@ -1,94 +1,99 @@
 
 public class PathSeqFilterSparkIntegrationTest extends CommandLineProgramTest {
 
+    static final String imagePath = "src/test/resources/" + PathSeqFilterSpark.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.img";
```
I believe there is a constant for "src/test/resouces" perhaps something like ```publicTestResourcesDir``` (yes in lowercase :( )

vruano replied:
Not sure of the name but I'm quite sure there is one.

mwalker174 replied:
Done

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSparkIntegrationTest.java:
```diff
@@ -1,94 +1,99 @@
+    @DataProvider(name = "pathseqFilterTestData")
+    public Object[][] getTestData() {
+        return new Object[][]{
+                {"basic_input.bam",
```
Is this the default indentation in IntellJ, it does not look great.

mwalker174 replied:
Yes it is. I removed one tab, which looks slightly better...

vruano commented:
@mwalker174 I'm asking of a bit of a re-haul of the specialized logging so perhaps you wanna seat down next week and go thru the code in tandem if you want.

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilter.java:
```diff
@@ -289,34 +290,28 @@ private static GATKRead clearReadAlignment(final GATKRead read, final SAMFileHea
-            throw new GATKException("Could not open metrics file " + metricsFileUri, e);
-        }
+    public PSFilterMetrics getFilterMetrics() {
+        Utils.nonNull(metrics, "Attempted to get PSFilter metrics, but it is null");
```
This code is gone now but I use ISE in some of the new code

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterMetrics.java:
```diff
@@ -0,0 +1,64 @@
+public final class PSFilterMetrics extends MetricBase implements Serializable {
+
+    /** Number of input reads that are not secondary or supplementary alignments */
+    public Long PRIMARY_READS;
```
Just trying to follow naming conventions for metrics output. I will ask engine team about this.

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterMetrics.java:
```diff
@@ -0,0 +1,64 @@
+     * Computes number of reads filtered at each step and total final reads.
+     */
+    public void computeDerivedMetrics() {
+        Utils.nonNull(PRIMARY_READS, "Cannot compute metrics if PRIMARY_READS is null");
```
Moved this to the logger class and throws ISE.

mwalker174 commented:
Thank you @vruano for your diligent review. I've implemented logger classes to encapsulate the metrics classes. Unfortunately the metrics classes must remain public in order to write output using `MetricsUtils.saveMetrics()`, but at least the tools aren't using them directly.

There are two logging class groups - one for Filter and one Score. For Filter, there is an interface `PSFilterLogger` that is implemented by a file-logging class `PSFilterFileLogger` and a dummy class `PSFilterEmptyLogger` that does nothing. There are analogous classes for Score, but there is no Empty logger because it's not actually necessary. This adds a lot of new classes (maybe you can think of a better way) but usage has been greatly simplified.

As we discussed in person, I don't think there is a faster way to count the reads in Spark. If you wanted to count the reads as they pass through, you would have to use some kind of atomic type that would be slow. Also it may be impossible to account for cases when tasks fail and restart.

@lbergelson @droazen In this PR, I wanted to use htsjdk's MetricsFile and MetricBase classes for writing metrics to a file. I notice that these classes are mostly used for picard-related things. Is this the preferred way to do things? They do force you to expose public variables and also use an upper-case naming convention. On the other hand, they are somewhat convenient.

vruano commented:
Looking at MetricFile and with it heavy use of Reflexion looks a bit nasty, if there is a better alternative the better. I guess that a better solution would use annotations to allow one to customize output variable name... force one to have those not-so-good looking CAPITAL_FIELD_NAMES for the sake of it is harsh. 

Only if the use of MetricFile is an overkill would ask you to do your custom one if it can be done in a few lines of code.... probably not the case.

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLogger.java:
```diff
@@ -0,0 +1,40 @@
+    public void logReadsAfterDeduplication(final JavaRDD<GATKRead> reads) { metrics.READS_AFTER_DEDUPLICATION = reads.count(); }
+    public void logFinalPairedReads(final JavaRDD<GATKRead> reads) { metrics.FINAL_PAIRED_READS = reads.count(); }
+
+    public void writeFile() {
```
Changed to close()

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreLogger.java:
```diff
@@ -0,0 +1,14 @@
+public interface PSScoreLogger {
+
+    void logReadCounts(final JavaRDD<GATKRead> reads);
+    void writeFile();
```
Done

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterMetrics.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Number of input reads that are not secondary or supplementary alignments
+     */
+    public Long PRIMARY_READS;
```
I wanted to use null's so I could tell if they have been assigned a value or not. I realized that since this class is public, it should check if they are null in computeDerivedMetrics() (before it was in the `PSScoreFileLogger` class). 

Similarly, PSScoreMetrics checks that the values are not null. If they are, however, the program will not crash. Rather, the metrics file writer will simply write an empty string for null values. 

I think is is the right behavior to output an empty string rather than writing, for example, 0, which is ambiguous.

mwalker174 commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -1,82 +1,60 @@
-            logger.error("Could not open test file " + fastqFile.getAbsolutePath(), e);
-        }
-        throw new TestException("Invalid fastq test file " + fastqFile);
+    private static final String BWA_IMAGE_PATH = publicTestDir + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.img";
```
Which comment? I thought you wanted to change this to static?

vruano replied:
Don't know now the comments seem to be gone.... if hg19mini.fasta is a common test reference it makes sense to add a image for it wherever it is included in the test resources (eg ./src/test/resources/hg19mini.img) and not in a deep down subdirectory that applies only to the class tested.

mwalker174 commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSKmerUtilsTest.java:
```diff
@@ -41,7 +41,7 @@ public void testGetKmersFromReference() throws IOException {
         }
 
-        final String expectedFile = "src/test/resources/" + PathSeqBuildKmers.class.getPackage().getName().replace(".", "/") + "/hg19mini.hss";
+        final String expectedFile = publicTestDir + PathSeqBuildKmers.class.getPackage().getName().replace(".", "/") + "/hg19mini.hss";
```
I couldn't find the constant to get the path given the current package. I only see one in BaseTest to get up to the "tools" directory.

mwalker174 commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLoggerTest.java:
```diff
@@ -0,0 +1,63 @@
+        final File expectedOutputFile = createTempFile("expected_metrics", ".txt");
+        MetricsUtils.saveMetrics(expectedMetricsFile, expectedOutputFile.getAbsolutePath());
+
+        Assert.assertTrue(MetricsFile.areMetricsEqual(metricsOutputFile, expectedOutputFile));
```
Hard-coded the expected numbers

vruano replied:
I often don't do add hard-wired test but I add a completely separated alternative implementation that is often slower and simpler (e.g. brute force) thus less prone to bugs. It is unlikely that both implementations would be equally affected except for trivial cases that perhaps you want to address in a separate test method (e.g. "empty" inputs whatever that would be depending on the case).

mwalker174 commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -1,82 +1,60 @@
-        throw new TestException("Invalid fastq test file " + fastqFile);
+    private static final String BWA_IMAGE_PATH = publicTestDir + PSFilter.class.getPackage().getName().replace(".", "/") + "/hg19mini.fasta.img";
+
+    static final GATKRead fastqRecordToGATKRead(final FastqRecord rec) {
```
Changed this and all below to private.

mwalker174 commented:
@vruano Addressed your comments. Please review when you have the chance.

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLoggerTest.java:
```diff
@@ -0,0 +1,44 @@
+import java.io.File;
+import java.util.Collections;
+
+public class PSFilterFileLoggerTest  extends BaseTest {
```
After #3475, this should extend `GATKBaseTest`

magicDGS commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreFileLoggerTest.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class PSScoreFileLoggerTest extends BaseTest {
```
After #3475, this should extend `GATKBaseTest`

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLoggerTest.java:
```diff
@@ -13,7 +13,7 @@
 
 public class PSFilterFileLoggerTest  extends GATKBaseTest {
-
+ 
```
This is odd, why you need to do this to force a rebuild? perhaps you need to ask David R and co how to handle it in travis appropriately. In any case please revert this empty extra line.

mwalker174 replied:
Done

vruano commented:
Additional changes requested. Back to  @mwalker174

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqPipelineSpark.java:
```diff
@@ -131,7 +131,7 @@ protected void runTool(final JavaSparkContext ctx) {
         JavaRDD<GATKRead> pairedReads = result._1;
         JavaRDD<GATKRead> unpairedReads = result._2;
-        filterLogger.writeFile();
+        filterLogger.close();
```
Done

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterEmptyLogger.java:
```diff
@@ -14,6 +14,6 @@ public void logReadsAfterQualityFilter(final JavaRDD<GATKRead> reads) {}
     public void logReadsAfterDeduplication(final JavaRDD<GATKRead> reads) {}
     public void logFinalPairedReads(final JavaRDD<GATKRead> reads) {}
-    public void writeFile() {}
+    public void close() {}
```
Done

mwalker174 replied:
Also added @Override to the other functions since they are defined in the interface.

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLogger.java:
```diff
@@ -27,12 +27,7 @@ public PSFilterFileLogger(final MetricsFile<PSFilterMetrics, Long> metricsFile,
-                || metrics.READS_AFTER_DEDUPLICATION == null || metrics.FINAL_PAIRED_READS == null) {
-            throw new IllegalStateException("Cannot write metrics if primary, pre-aligned host, quality, host, duplicate, or final paired read counts are not logged");
-        }
+    public void close() {
```
Done

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterLogger.java:
```diff
@@ -14,6 +14,6 @@
     void logReadsAfterDeduplication(final JavaRDD<GATKRead> reads);
     void logFinalPairedReads(final JavaRDD<GATKRead> reads);
-    void writeFile();
+    void close();
```
Done

mwalker174 commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreFileLogger.java:
```diff
@@ -29,10 +29,7 @@ public void logReadCounts(final JavaRDD<GATKRead> reads) {
-        if (metrics.MAPPED_READS == null || metrics.UNMAPPED_READS == null) {
-            throw new IllegalStateException("Cannot write metrics if mapped or unmapped read counts are not logged");
-        }
+    public void close() {
```
Done

mwalker174 commented:
Thank you @vruano. I've addressed your comments to implement AutoCloseable in the loggers and use try-with-resources. I also moved the hg19mini BWA image file to the base resources directory.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java:
```diff
@@ -99,6 +109,7 @@ protected void runTool(final JavaSparkContext ctx) {
         } else {
             logger.info("No unpaired reads to write - BAM will not be written.");
         }
+
```
Remove blank line (if need to change this file for something else, otherwise forget it).

mwalker174 replied:
No changes to this file

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterEmptyLogger.java:
```diff
@@ -0,0 +1,26 @@
+ */
+public final class PSFilterEmptyLogger implements PSFilterLogger {
+
+    @Override
```
One blank lines between method declarations.

mwalker174 replied:
Done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLogger.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public void logPrimaryReads(final JavaRDD<GATKRead> reads) { metrics.PRIMARY_READS = reads.count(); }
+    public void logReadsAfterPrealignedHostFilter(final JavaRDD<GATKRead> reads) { metrics.READS_AFTER_PREALIGNED_HOST_FILTER = reads.count(); }
```
Missing [at]Override.
Also please use 3 lines for each method.

mwalker174 replied:
Done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterFileLogger.java:
```diff
@@ -0,0 +1,41 @@
+    private final String metricsOutputPath;
+
+    public PSFilterFileLogger(final MetricsFile<PSFilterMetrics, Long> metricsFile, final String metricsOutputPath) {
+        this.metrics = new PSFilterMetrics();
```
Do you need to test for nulls here?

mwalker174 replied:
Not a bad idea

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSFilterLogger.java:
```diff
@@ -0,0 +1,20 @@
+    void logReadsAfterDeduplication(final JavaRDD<GATKRead> reads);
+    void logFinalPairedReads(final JavaRDD<GATKRead> reads);
+    @Override
+    void close();
```
There is no need to redeclare a super-interface method here (```close```).

mwalker174 replied:
Done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreFileLogger.java:
```diff
@@ -0,0 +1,38 @@
+    private final String metricsOutputPath;
+
+    public PSScoreFileLogger(final MetricsFile<PSScoreMetrics, Long> metricsFile, final String metricsOutputPath) {
+        this.metrics = new PSScoreMetrics();
```
Check for nulls.

mwalker174 replied:
Done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreFileLogger.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void logReadCounts(final JavaRDD<GATKRead> reads) {
+        final long numReads = reads.count();
```
check for nulls.

mwalker174 replied:
Done

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/loggers/PSScoreLogger.java:
```diff
@@ -0,0 +1,15 @@
+
+    void logReadCounts(final JavaRDD<GATKRead> reads);
+    @Override
+    void close();
```
Remove redundant re-declaration.

mwalker174 replied:
Done

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -1,88 +1,66 @@
-            return read;
-        }).collect(Collectors.toList());
+    private static final List<GATKRead> getReadsFromFastq(final File fastqFile) {
+        final FastqReader reader = new FastqReader(fastqFile);
```
I think you need to close the reader.... I guess you should use a try-with-resources.

mwalker174 replied:
Done

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -1,88 +1,66 @@
-            read.setIsPaired(true);
-            read.setIsSecondOfPair();
-        });
+        for (GATKRead gatkRead : readList1) {
```
You can make ```gatkRead``` final.

mwalker174 replied:
Done

vruano commented on src/test/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSFilterTest.java:
```diff
@@ -1,88 +1,66 @@
+            gatkRead.setIsPaired(true);
+            gatkRead.setIsFirstOfPair();
+        }
+        for (GATKRead gatkRead : readList2) {
```
here too.

mwalker174 replied:
Done
